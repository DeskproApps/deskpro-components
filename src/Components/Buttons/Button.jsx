import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import noop from 'utils/noop';
import { objectKeyFilter } from 'utils/objects';

/**
 * Standard button.
 */
export default class Button extends React.Component {
  static propTypes = {
    /**
     * Displays the button at the given size
     */
    size:                    PropTypes.oneOf(['s', 'm', 'l', 'small', 'medium', 'large']),
    /**
     * The type of button to display.
     */
    type:                    PropTypes.oneOf(['primary', 'secondary', 'cta']),
    /**
     * The shape of the button.
     */
    shape:                   PropTypes.oneOf(['default', 'square', 'round']),
    /**
     * Children to render.
     */
    children:                PropTypes.node.isRequired,
    /**
     * CSS classes to apply to the element.
     */
    className:               PropTypes.string,
    /**
     * Whether or not the button is disabled or not.
     */
    disabled:                PropTypes.bool,
    /**
     * Called when the user clicks outside of the button.
     */
    onClickOutside:          PropTypes.func,
    /**
     * Disables outside click listening by explicitly removing the event listening bindings.
     */
    disableOnClickOutside:   PropTypes.func,
    /**
     * Enables outside click listening by setting up the event listening bindings.
     */
    enableOnClickOutside:    PropTypes.func,
    /**
     * Whether the event should be propagated
     */
    stopPropagation:         PropTypes.bool,
    /**
     * Whether the event should prevent default behaviour
     */
    preventDefault:          PropTypes.bool,
    /**
     * Property passed by OnClickOutside HOC
     */
    eventTypes:              PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
    /**
     * Property passed by OnClickOutside HOC
     */
    outsideClickIgnoreClass: PropTypes.string,
  };

  static defaultProps = {
    size:                    'large',
    className:               '',
    type:                    'primary',
    shape:                   'default',
    disabled:                false,
    onClickOutside:          noop,
    disableOnClickOutside:   noop,
    enableOnClickOutside:    noop,
    stopPropagation:         false,
    preventDefault:          false,
    eventTypes:              ['mousedown', 'touchstart'],
    outsideClickIgnoreClass: '',
  };

  handleClickOutside = (event) => {
    this.props.onClickOutside(event);
  };

  render() {
    const { size, type, shape, disabled, className, children, ...props } = this.props;

    return (
      <button
        disabled={disabled}
        className={classNames(
          'dp-button',
          `dp-button--${size[0]}`,
          `dp-button--${type}`,
          `dp-button--shape-${shape}`,
          className
        )}
        {...objectKeyFilter(props, Button.propTypes)}
      >
        {children}
      </button>
    );
  }
}
