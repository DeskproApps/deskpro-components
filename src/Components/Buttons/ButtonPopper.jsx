import React from 'react';

/**
 * Container for a popper which is attached to a button.
 */
const ButtonPopper = props => (
  <span {...props} />
);

export default ButtonPopper;
