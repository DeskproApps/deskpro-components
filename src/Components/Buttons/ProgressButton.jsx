import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { objectKeyFilter } from 'utils/objects';
import Progress from 'Components/Progress';
import ProgressBar from 'Components/ProgressBar';

/**
 * Renders a button which contains a progress bar.
 */
export default class ProgressButton extends React.Component {
  static propTypes = {
    /**
     * The percentage complete.
     */
    percent:   PropTypes.number,
    /**
     * The button size.
     */
    size:      PropTypes.oneOf(['s', 'm', 'l', 'small', 'medium', 'large']),
    /**
     * CSS classes to apply to the element.
     */
    className: PropTypes.string,
    /**
     * Children to render.
     */
    children:  PropTypes.node
  };

  static defaultProps = {
    percent:   0,
    size:      'large',
    className: '',
    children:  ''
  };

  render() {
    const {
      percent, size, className, children, ...props
    } = this.props;

    return (
      <div
        className={classNames('dp-progress-button', `dp-progress-button--${size[0]}`, className)}
        {...objectKeyFilter(props, ProgressButton.propTypes)}
      >
        <div className="dp-progress-button__label">
          {children}
        </div>
        <Progress size="small" type={percent < 100 ? 'primary' : 'cta'}>
          <ProgressBar percent={percent} />
        </Progress>
      </div>
    );
  }
}
