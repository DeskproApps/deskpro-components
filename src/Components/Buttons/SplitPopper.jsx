import React from 'react';

/**
 * Container for the popper contents of a split button
 */
const SplitPopper = props => (
  <span {...props} />
);

export default SplitPopper;
