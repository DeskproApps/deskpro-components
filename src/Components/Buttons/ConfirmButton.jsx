import React from 'react';
import PropTypes from 'prop-types';
import onClickOutside from 'react-onclickoutside';
import noop from '../../utils/noop';
import { objectKeyFilter } from '../../utils/objects';
import Button from './Button';

const ClickOutsideButton = onClickOutside(Button);

/**
 * Button with click confirmation.
 */
export default class ConfirmButton extends React.Component {
  static propTypes = {
    /**
     * Whether the component is disabled.
     */
    disabled:              PropTypes.bool,
    /**
     * The confirmation message.
     */
    message:               PropTypes.string,
    /**
     * Children to render.
     */
    children:              PropTypes.node.isRequired,
    /**
     * Disables outside click listening by explicitly removing the event listening bindings.
     */
    disableOnClickOutside: PropTypes.func,
    /**
     * Enables outside click listening by setting up the event listening bindings.
     */
    enableOnClickOutside:  PropTypes.func,
    /**
     * Called when the button is clicked.
     */
    onClick:               PropTypes.func
  };

  static defaultProps = {
    disabled:              false,
    message:               'Are you sure?',
    onClick:               noop,
    disableOnClickOutside: noop,
    enableOnClickOutside:  noop,
  };

  constructor(props) {
    super(props);
    this.state = {
      confirm: false
    };
  }

  getLabel = () => {
    if (this.state.confirm) {
      return this.props.message;
    }
    return this.props.children;
  };

  handleClick = (e) => {
    const { disabled } = this.props;
    e.preventDefault();
    if (!disabled) {
      if (this.state.confirm) {
        this.props.onClick(e);
        this.setState({
          confirm: false
        });
      } else {
        this.setState({
          confirm: true
        });
      }
    }
  };

  clickOutside = () => {
    this.setState({
      confirm: false
    });
  };

  render() {
    return (
      <ClickOutsideButton
        onClick={this.handleClick}
        onClickOutside={this.clickOutside}
        disableOnClickOutside={!this.state.confirm}
        disabled={this.props.disabled}
        {...objectKeyFilter(this.props, ConfirmButton.propTypes)}
      >
        {this.getLabel()}
      </ClickOutsideButton>
    );
  }
}

