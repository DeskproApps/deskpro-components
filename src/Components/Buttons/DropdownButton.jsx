import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { faCaretDown } from '@fortawesome/free-solid-svg-icons';
import noop from '../../utils/noop';
import { objectKeyFilter } from '../../utils/objects';
import { cssMatchComputedWidth } from '../../utils/css';
import Button from './Button';
import ButtonPopper from './ButtonPopper';
import { Popper } from '../Common';
import Icon from '../Icon';

/**
 * Renders a button with a drop down icon and popper.
 */
export default class DropdownButton extends React.Component {
  static propTypes = {
    /**
     * The size of the button
     */
    size:      PropTypes.string,
    /**
     * The type of the button
     */
    type:      PropTypes.string,
    /**
     * CSS classes to apply to the element.
     */
    className: PropTypes.string,
    /**
     * The component styles.
     */
    style:     PropTypes.object,
    /**
     * Children to render.
     */
    children:  PropTypes.node,
    /**
     * Called when the button is clicked.
     */
    onClick:   PropTypes.func
  };

  static defaultProps = {
    size:      'large',
    type:      'primary',
    onClick:   noop,
    className: '',
    children:  [],
    style:     {}
  };

  constructor(props) {
    super(props);
    this.rootRef   = null;
    this.rootDOM   = null;
    this.popperRef = null;
    this.popperDOM = null;
  }

  componentDidMount() {
    this.rootDOM   = ReactDOM.findDOMNode(this.rootRef);
    this.popperDOM = ReactDOM.findDOMNode(this.popperRef);
  }

  /**
   * Sets the width of the popper to match the button width
   */
  updatePopperWidth = () => {
    cssMatchComputedWidth(this.rootRef, this.popperRef);
  };

  handleClick = (e) => {
    this.popperRef.toggle();
    this.props.onClick(e);
  };

  handleOpen = () => {
    this.updatePopperWidth();
  };

  render() {
    const { className, children, ...props } = this.props;

    const buttonChildren = [];
    let popperChild    = <ButtonPopper />;
    React.Children.toArray(children).forEach((child) => {
      if (child.type === ButtonPopper) {
        popperChild = child;
      } else {
        buttonChildren.push(child);
      }
    });

    return (
      <Button
        ref={ref => (this.rootRef = ref)}
        onClick={this.handleClick}
        className={classNames('dp-dropdown-button', className)}
        {...objectKeyFilter(props, DropdownButton.propTypes)}
      >
        {buttonChildren}
        <Icon name={faCaretDown} />
        <Popper
          placement="bottom-end"
          ref={ref => (this.popperRef = ref)}
          target={this.rootRef}
          arrow={false}
          onOpen={this.handleOpen}
          {...popperChild.props}
        >
          {popperChild}
        </Popper>
      </Button>
    );
  }
}
