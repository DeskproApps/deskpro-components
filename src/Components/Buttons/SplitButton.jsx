import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import noop from 'utils/noop';
import { objectKeyFilter } from 'utils/objects';
import { cssMatchComputedWidth } from 'utils/css';
import Button from 'Components/Buttons/Button';
import ButtonPopper from 'Components/Buttons/ButtonPopper';
import Icon from 'Components/Icon';
import { Popper } from 'Components/Common';

/**
 * Renders a button with a left side and right side, where clicking the right side opens
 * a popper.
 */
export default class SplitButton extends React.Component {
  static propTypes = {
    /**
     * Displays the button at the given size
     */
    size:      PropTypes.oneOf(['s', 'm', 'l', 'small', 'medium', 'large']),
    /**
     * The type of button to display.
     */
    type:      PropTypes.oneOf(['primary', 'secondary', 'cta']),
    /**
     * Whether or not the button is disabled or not.
     */
    disabled:  PropTypes.bool,
    /**
     * CSS classes to apply to the element.
     */
    className: PropTypes.string,
    /**
     * Styles added to the element.
     */
    style:     PropTypes.object,
    /**
     * Children to render.
     */
    children:  PropTypes.node,
    /**
     * Called when the left side of the button is clicked.
     */
    onClick:   PropTypes.func
  };

  static defaultProps = {
    size:      'large',
    type:      'primary',
    disabled:  false,
    onClick:   noop,
    children:  [],
    className: '',
    style:     {}
  };

  constructor(props) {
    super(props);
    this.rootRef   = null;
    this.rootDOM   = null;
    this.popperRef = null;
    this.popperDOM = null;
  }

  componentDidMount() {
    this.rootDOM   = ReactDOM.findDOMNode(this.rootRef);
    this.popperDOM = ReactDOM.findDOMNode(this.popperRef);
  }

  /**
   * Sets the width of the popper to match the button width
   */
  updatePopperWidth = () => {
    cssMatchComputedWidth(this.rootDOM, this.popperDOM);
  };

  handleRightClick = () => {
    this.popperRef.toggle();
  };

  handleOpen = () => {
    this.updatePopperWidth();
  };

  render() {
    const {
      size, type, disabled, onClick, className, children, ...props
    } = this.props;

    const leftChildren = [];
    let popperChild    = <ButtonPopper />;
    React.Children.toArray(children).forEach((child) => {
      if (child.type === ButtonPopper) {
        popperChild = child;
      } else {
        leftChildren.push(child);
      }
    });

    return (
      <div
        ref={ref => (this.rootRef = ref)}
        className={classNames('dp-split-button', className)}
        {...objectKeyFilter(props, SplitButton.propTypes)}
      >
        <Button
          className="dp-split-button__left"
          size={size}
          type={type}
          disabled={disabled}
          onClick={onClick}
        >
          {leftChildren}
        </Button>
        <Button
          className="dp-split-button__right"
          size={size}
          type={type}
          disabled={disabled}
          onClick={this.handleRightClick}
        >
          <Icon name="caret-down" />
        </Button>
        <Popper
          placement="bottom-end"
          ref={ref => (this.popperRef = ref)}
          target={this.rootRef}
          arrow={false}
          onOpen={this.handleOpen}
          {...popperChild.props}
        >
          {popperChild}
        </Popper>
      </div>
    );
  }
}
