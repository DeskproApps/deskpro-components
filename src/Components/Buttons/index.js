import Button from './Button';
import ConfirmButton from './ConfirmButton';
import SplitButton from './SplitButton';
import ButtonPopper from './ButtonPopper';
import DropdownButton from './DropdownButton';
import ProgressButton from './ProgressButton';
import TranslateButton from './TranslateButton';

export {
  Button,
  ConfirmButton,
  SplitButton,
  ButtonPopper,
  DropdownButton,
  ProgressButton,
  TranslateButton,
};
