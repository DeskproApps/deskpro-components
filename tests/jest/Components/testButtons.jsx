import React from 'react';
import renderer from 'react-test-renderer';
import { Button, ButtonPopper, ProgressButton, TranslateButton, DropdownButton } from 'Components/Buttons/index';

// https://github.com/facebook/react/issues/7740#issuecomment-247335106
const createNodeMock = () => ({
  addEventListener: () => {}
});

describe('>>> Buttons --- Snapshot', () => {
  it('+++capturing Snapshot of Button', () => {
    const renderedValue = renderer.create(
      <Button
        size="small"
        type="primary"
      >
        Button
      </Button>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
  it('+++capturing Snapshot of ButtonPopper', () => {
    const renderedValue = renderer.create(
      <ButtonPopper
        size="small"
        type="primary"
      >
        Button
      </ButtonPopper>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
  it('+++capturing Snapshot of ProgressButton', () => {
    const renderedValue = renderer.create(
      <ProgressButton percent={75} size="large">
        75
      </ProgressButton>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
  it('+++capturing Snapshot of TranslateButton', () => {
    const renderedValue = renderer.create(
      <TranslateButton percent={65} size="l">
        65
      </TranslateButton>
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
  it('+++capturing Snapshot of DropdownButton', () => {
    const renderedValue = renderer.create(
      <DropdownButton size="l" type="primary">
        Test
      </DropdownButton>,
      { createNodeMock }
    ).toJSON();
    expect(renderedValue).toMatchSnapshot();
  });
});
