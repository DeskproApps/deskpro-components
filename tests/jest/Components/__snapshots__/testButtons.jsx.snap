// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Buttons --- Snapshot +++capturing Snapshot of Button 1`] = `
<button
  className="dp-button dp-button--s dp-button--primary dp-button--shape-default"
  disabled={false}
>
  Button
</button>
`;

exports[`>>> Buttons --- Snapshot +++capturing Snapshot of ButtonPopper 1`] = `
<span
  size="small"
  type="primary"
>
  Button
</span>
`;

exports[`>>> Buttons --- Snapshot +++capturing Snapshot of DropdownButton 1`] = `
<button
  className="dp-button dp-button--l dp-button--primary dp-button--shape-default dp-dropdown-button"
  disabled={false}
  onClick={[Function]}
>
  Test
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-caret-down fa-w-10 fa-sm dp-icon"
    data-icon="caret-down"
    data-prefix="fas"
    role="img"
    style={Object {}}
    viewBox="0 0 320 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <div />
</button>
`;

exports[`>>> Buttons --- Snapshot +++capturing Snapshot of ProgressButton 1`] = `
<div
  className="dp-progress-button dp-progress-button--l"
>
  <div
    className="dp-progress-button__label"
  >
    75
  </div>
  <div
    className="dp-progress dp-progress--primary dp-progress--s"
  >
    <div
      className="dp-progress__bar"
      style={
        Object {
          "width": "75%",
        }
      }
      type="primary"
    >
      
    </div>
  </div>
</div>
`;

exports[`>>> Buttons --- Snapshot +++capturing Snapshot of SplitButton 1`] = `
<div
  className="dp-split-button"
>
  <button
    className="dp-button dp-button--l dp-button--primary dp-button--shape-default dp-split-button__left"
    disabled={false}
    onClick={[Function]}
  >
    Test
  </button>
  <button
    className="dp-button dp-button--l dp-button--primary dp-button--shape-default dp-split-button__right"
    disabled={false}
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      className="svg-inline--fa fa-caret-down fa-w-10 fa-sm dp-icon"
      data-icon="caret-down"
      data-prefix="fas"
      role="img"
      style={Object {}}
      viewBox="0 0 320 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
        fill="currentColor"
        style={Object {}}
      />
    </svg>
  </button>
  <div />
</div>
`;

exports[`>>> Buttons --- Snapshot +++capturing Snapshot of TranslateButton 1`] = `
<div
  className="dp-translate-button dp-translate-button--l"
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-globe fa-w-16 fa-sm dp-icon"
    data-icon="globe"
    data-prefix="fas"
    role="img"
    style={Object {}}
    viewBox="0 0 496 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M336.5 160C322 70.7 287.8 8 248 8s-74 62.7-88.5 152h177zM152 256c0 22.2 1.2 43.5 3.3 64h185.3c2.1-20.5 3.3-41.8 3.3-64s-1.2-43.5-3.3-64H155.3c-2.1 20.5-3.3 41.8-3.3 64zm324.7-96c-28.6-67.9-86.5-120.4-158-141.6 24.4 33.8 41.2 84.7 50 141.6h108zM177.2 18.4C105.8 39.6 47.8 92.1 19.3 160h108c8.7-56.9 25.5-107.8 49.9-141.6zM487.4 192H372.7c2.1 21 3.3 42.5 3.3 64s-1.2 43-3.3 64h114.6c5.5-20.5 8.6-41.8 8.6-64s-3.1-43.5-8.5-64zM120 256c0-21.5 1.2-43 3.3-64H8.6C3.2 212.5 0 233.8 0 256s3.2 43.5 8.6 64h114.6c-2-21-3.2-42.5-3.2-64zm39.5 96c14.5 89.3 48.7 152 88.5 152s74-62.7 88.5-152h-177zm159.3 141.6c71.4-21.2 129.4-73.7 158-141.6h-108c-8.8 56.9-25.6 107.8-50 141.6zM19.3 352c28.6 67.9 86.5 120.4 158 141.6-24.4-33.8-41.2-84.7-50-141.6h-108z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <div
    className="dp-progress-button dp-progress-button--l"
  >
    <div
      className="dp-progress-button__label"
    >
      65
    </div>
    <div
      className="dp-progress dp-progress--primary dp-progress--s"
    >
      <div
        className="dp-progress__bar"
        style={
          Object {
            "width": "65%",
          }
        }
        type="primary"
      >
        
      </div>
    </div>
  </div>
</div>
`;
